lib/netif.ml

```diff
@@ -361,105 +444,100 @@ let connect id =
    to Xenstore? XXX *)
 let disconnect t =
   printf "Netif: disconnect\n%!";
+  Shared_page_pool.shutdown t.t.tx_pool;
   Hashtbl.remove devices t.t.id;
   return ()
 
-let page_size = Io_page.round_to_page_size 1
-
-(* Push a single page to the ring, but no event notification *)
-let write_request ?size ~flags nf page =
-  let len = Cstruct.len page in
-  if page.Cstruct.off + len > page_size then begin
-    (* netback rejects packets that cross page boundaries *)
-    let msg =
-      Printf.sprintf "Invalid page: offset=%d, length=%d" page.Cstruct.off len in
-    print_endline msg;
-    Lwt.fail (Failure msg)
-  end else
-  lwt gref = Gnt.Gntshr.get () in
-  (* This grants access to the *base* data pointer of the page *)
-  (* XXX: another place where we peek inside the cstruct *)
-  Gnt.Gntshr.grant_access ~domid:nf.t.backend_id ~writable:false gref page.Cstruct.buffer;
-  let size = match size with |None -> len |Some s -> s in
-  (* XXX: another place where we peek inside the cstruct *)
-  nf.t.stats.tx_pkts <- Int32.succ nf.t.stats.tx_pkts;
-  nf.t.stats.tx_bytes <- Int64.add nf.t.stats.tx_bytes (Int64.of_int size);
-  let offset = page.Cstruct.off in
-  lwt replied = Lwt_ring.Front.write nf.t.tx_client
-      (TX.Proto_64.write ~id:gref ~gref:(Int32.of_int gref) ~offset ~flags ~size) in
-  (* request has been written; when replied returns we have a reply *)
-  let replied =
-    try_lwt
-      lwt _ = replied in
-      Gnt.Gntshr.end_access gref;
-      Gnt.Gntshr.put gref;
-      return ()
-    with Lwt_ring.Shutdown ->
-      Gnt.Gntshr.put gref;
-      fail Lwt_ring.Shutdown
-       | e ->
-         Gnt.Gntshr.end_access gref;
-         Gnt.Gntshr.put gref;
-         fail e in
-  return replied
-
-(* Transmit a packet from buffer, with offset and length *)  
-let rec write_already_locked nf page =
-  try_lwt
-    lwt th = write_request ~flags:0 nf page in
-    Lwt_ring.Front.push nf.t.tx_client (notify nf.t);
-    lwt () = th in
-    (* all fragments acknowledged, resources cleaned up *)
-    return ()
-  with | Lwt_ring.Shutdown -> write_already_locked nf page
-
-let write nf page =
-  Lwt_mutex.with_lock nf.t.tx_mutex
-    (fun () ->
-       write_already_locked nf page
-    )
+(** Copy from src to dst until src is exhausted or dst is full.
+ * Returns the number of bytes copied and the remaining data from src, if any. *)
+let blitv src dst =
+  let rec aux dst n = function
+    | [] -> n, []
+    | hd::tl ->
+        let avail = Cstruct.len dst in
+        let first = Cstruct.len hd in
+        if first <= avail then (
+          Cstruct.blit hd 0 dst 0 first;
+          aux (Cstruct.shift dst first) (n + first) tl
+        ) else (
+          Cstruct.blit hd 0 dst 0 avail;
+          let rest_hd = Cstruct.shift hd first in
+          (n + avail, rest_hd :: tl)
+        ) in
+  aux dst 0 src
```

Comment:
could be useful to upstream his into Cstruct (more generally we need more Cstruct functions working on list of Cstructs I think)